cli/CommandExecutor.cpp

```diff
@@ -0,0 +1,189 @@
+  } else if (command_str.compare("\\d") == 0) {
+    if (arguments->size() == 0) {
+      executeDescribeDatabase(arguments, catalog_database, out);
+    } else {
+      executeDescribeTable(arguments, catalog_database, out);
+    }
+  } else {
+    THROW_SQL_ERROR_AT(command.command()) << "Invalid Command";
+  }
+}
+
+void CommandExecutor::executeDescribeDatabase(
+    const PtrVector<ParseString> *arguments,
+    const CatalogDatabase *catalog_database, FILE *out) {
+  std::size_t relation_count = 0;
+  // Column width initialized to 6 to take into account the header name
+  // and the column value table
+  std::size_t max_column_width = CommandExecutor::kInitMaxColumnWidth;
+  const CatalogRelation *relation;
+  if (arguments->size() == 0) {
+    for (auto rel_it = catalog_database->begin();
+         rel_it != catalog_database->end(); ++rel_it) {
+      std::size_t column_width = rel_it->getName().length();
+      max_column_width =
+          max_column_width < column_width ? column_width : max_column_width;
+      relation_count++;
+    }
+  } else {
+    const ParseString &table_name = arguments->front();
+    const std::string &table_name_val = table_name.value();
+    relation = catalog_database->getRelationByName(table_name_val);
+    std::size_t column_width = relation->getName().length();
+    max_column_width =
+        max_column_width < column_width ? column_width : max_column_width;
+    ++relation_count;
+  }
+  // Only if we have relations work on the printing logic.
+  if (relation_count > 0) {
+    vector<std::size_t> column_widths;
+    column_widths.push_back(max_column_width);
+    column_widths.push_back(CommandExecutor::kInitMaxColumnWidth);
+    fputs("       List of relations\n\n", out);
+    fprintf(out, "%-*s|", static_cast<int>(max_column_width)+1, " Name");
+    fprintf(out, "%-*s\n", static_cast<int>(CommandExecutor::kInitMaxColumnWidth), " Type");
+    PrintToScreen::printHBar(column_widths, out);
+    //  If there are no argument print the entire list of tables
+    //  else print the particular table only.
+    if (arguments->size() == 0) {
+      for (auto rel_it = catalog_database->begin();
+           rel_it != catalog_database->end(); ++rel_it) {
+        fprintf(out, " %-*s|", static_cast<int>(max_column_width), rel_it->getName().c_str());
+        fprintf(out, " %-*s\n", static_cast<int>(CommandExecutor::kInitMaxColumnWidth), "table");
+      }
+    } else {
+      fprintf(out, " %-*s|", static_cast<int>(max_column_width), relation->getName().c_str());
+      fprintf(out, " %-*s\n", static_cast<int>(CommandExecutor::kInitMaxColumnWidth), "table");
+    }
+    fputc('\n', out);
+  }
+}
+
+void CommandExecutor::executeDescribeTable(
+    const PtrVector<ParseString> *arguments,
+    const CatalogDatabase *catalog_database, FILE *out) {
+  const ParseString &table_name = arguments->front();
+  const std::string &table_name_val = table_name.value();
+  const CatalogRelation *relation =
```

Comment:
Careful here. Unless I missed a check elsewhere in the code, `getRelationByName` will return NULL if a relation does not exist by that name.